feat: IPNS Publishing

hacdias opened the pull request:
Closes #1482.

-  [x] Add IPNS key management UI to Settings page
-  [x] Publishing to IPNS id via context action on Files page

## A bunch of screenshots

![Screen Shot 2021-09-24 at 13 29 10](https://user-images.githubusercontent.com/5447088/134667351-905fb984-296b-4234-a981-a65bc53f948d.png)

The context menu is a bit unaligned. The same happens with the Pinning Services context menu. The target reference seems correctly set so I'm not sure why's this miscalculation happening.

![Screen Shot 2021-09-24 at 13 29 27](https://user-images.githubusercontent.com/5447088/134667380-eb6da5bd-7455-4daa-b436-6076e6b757ba.png)

![Screen Shot 2021-09-24 at 13 37 16](https://user-images.githubusercontent.com/5447088/134668211-079ad0fd-f090-47e4-ba66-82a30ba43687.png)

![Screen Shot 2021-09-24 at 13 37 27](https://user-images.githubusercontent.com/5447088/134668233-63a09718-5ded-414c-a3f6-d443d54abfb3.png)

![Screen Shot 2021-09-24 at 13 37 35](https://user-images.githubusercontent.com/5447088/134668242-0198cc97-7b03-4558-bd45-ad29e49feb7a.png)

![Screen Shot 2021-09-24 at 13 37 50](https://user-images.githubusercontent.com/5447088/134668262-80403c4d-63f9-41eb-9ca2-531f87e78285.png)

![image](https://user-images.githubusercontent.com/5447088/134668399-e799b7c6-59ef-4104-bbc7-b0599134d838.png)

hacdias commented:
This one is ready @lidel :D

lidel commented on src/bundles/ipns.js:
```diff
@@ -0,0 +1,63 @@
+  doGenerateIpnsKey: (name) => async ({ getIpfs, store }) => {
+    const ipfs = getIpfs()
+
+    // Stronger defaults than the listed on the docs:
+    // https://github.com/ipfs/js-ipfs/blob/master/docs/core-api/KEY.md#ipfskeygenname-options
+    await ipfs.key.gen(name, {
+      type: 'ed25519',
+      size: 4096
+    })
```
I believe those defaults changed a while ago in both js-ipfs and go-ipfs.
- [ ] TODO: check if we can switch to the default now.

hacdias replied:
Ah, both `js-ipfs` and `go-ipfs` default to keys of type `ed25519`. As far as I understand, the size does not influence this types of keys, so I will clean that up.

lidel reviewed (commented):
Finally, got to spend some time with this :)
Works really good, want to include it in go-ipfs 0.11 if possible.

Before we merge this, need to add some e2e tests tomorrow:
- [ ] add key and confirm pet name is added to the table
  - [ ] publish to the key from files
    - [ ] confirm link on the settings screen got updated 
  - [ ] remove key and confirm it's gone from the table


For now, pushed small changes which tweak  some modals a bit:


> ![Screenshot 2021-11-25 at 02-07-29  animals Files IPFS](https://user-images.githubusercontent.com/157609/143334364-6a620ee5-a1d2-49f1-ad9a-036ced1d7176.png)

> ![Screenshot 2021-11-25 at 02-10-31 Settings IPFS](https://user-images.githubusercontent.com/157609/143334355-8ffc7003-db85-4413-a8e0-c80356e07ff7.png)

> ![Screenshot 2021-11-25 at 02-10-00 Settings IPFS](https://user-images.githubusercontent.com/157609/143334335-cb23c481-1f11-4d76-84b8-4e167c924637.png)

lidel reviewed (changes requested):
Added E2E tests and cleaned them up a bit, should  be pretty robust now that we use playwright. 

--- 

@hacdias I think we have UX gap around publishing, which blocks this from being shipped.

Need some sort of confirmation that publishing was successful. 

## Visual feedback on IPN Publish

Right now I click Publish and nothing happens. 
We need three states:
- "Publishing in progress"
- "Publishing successful" / "Publishing failed"

A nice flow would be if clicking on "Publish"  displayed spinner until `ipfs.name.publish` returns result (success/error), and then:
- if error, show error inside of the modal.
- if success, show "Succes, published at {keyid}" and below the same UI we have on "Share" modal, where user can copy a shareable link to `/ipns/` address at a public gateway.


@hacdias will you have bandwidth to add this? 
I think it is better to wait and polish the experience, so I'm descoping this from go-ipfs 0.11, so there is no rush.

hacdias commented:
@lidel I'll work on it.

hacdias commented:
@lidel it's still missing the actual spinner. Could you just validate if the workflow is good?

lidel commented:
@hacdias yes, workflow is :ok_hand:  

Figuring out an engaging spinner will be tricky – some thoughts:
- IPNS publishing can take up to a minute atm because (iirc) it puts IPNS record in 20 peers, and it has to find the right ones first.  (It is faster if [Accelerated DHT Client](https://github.com/ipfs/go-ipfs/blob/master/docs/experimental-features.md#accelerated-dht-client) is enabled, but can't use it by default yet).
- A fake progressbar that assumes to take ~1m should be enough of feedback / setting expectations. We can improve it later.
- If you want to get creative, we could inform user about DHT traversal, eg. show random DHT walk with  [await first(ipfs.log.tail())](https://github.com/ipfs/js-ipfs/blob/484d2559cfe120f01eb9c5dc728d1f7cfbd89730/packages/ipfs-http-client/test/log.spec.js#L21-L32) or fake it.

ps. fysa pushed two changes:
- Finished E2E tests for IPNS publishing – they now confirm IPNS record was updated by doing `ipfsn.name.resolve`.
- Added some copy text above the shareable link. (just a draft, when we are done we will ask a native speaker to proofread and refine messaging and UI)
  >  ![Screenshot 2021-12-03 at 00-35-33  animals Files IPFS](https://user-images.githubusercontent.com/157609/144520112-6195c37e-62a9-48c1-a58b-9760090fe876.png)

hacdias commented:
> A fake progressbar that assumes to take ~1m should be enough of feedback / setting expectations. We can improve it later.

I will do this. Seems to be the easiest. Using a spinner would probably not be the greatest idea idea, indeed, as it does not give feedback.

hacdias commented:
@lidel I added a progress bar that is initially 60 seconds. After each publishing, we average the current expected time with the new time it take. I think that this way, it should improve and have better expectations over time.

hacdias commented:
Ping @lidel

lidel commented on src/files/modals/publish-modal/PublishModal.js:
```diff
@@ -0,0 +1,157 @@
+  useEffect(() => {
+    if (!start) return
+
+    const interval = setInterval(() => {
```
:thought_balloon:  nit: I wonder if we could remove `setInterval` here and use [css animations](https://stackoverflow.com/question/15945288/animating-progress-bars-with-css) instead somehow?  (would look less choppy, we could also make it more engaging by making it non-linear). 

I'm thinking setting `progress` to 100% and using  `animation: progressBar ${expectedPublishTime}s ease-in-out`  for customizing the progress animation instead.

Just an idea, feel free to ignore.

lidel reviewed (changes requested):
@hacdias really nice! 

* I adjusted the rolling average to overshoot rather than undershoot (apps fake progress and usually finish around 75% to give that "snappy feeling", we guess based on past times, which is better, but still should overshoot to avoid being stuck at 100%) + added brief explainer what is happening, so the wait is a bit more justified:
  > ![2021-12-17_23-47](https://user-images.githubusercontent.com/157609/146616734-94f677f0-ba9f-4312-98a1-a018654073f1.png)
- I'm marking this as blocked on go-ipfs roadmap:
  - We need to figure out  if it is ok to ship the GUI for IPNS before we have things like  https://github.com/ipfs/go-ipfs/issues/8591 and https://github.com/ipfs/go-ipfs/issues/8586 in go-ipfs, or are those worth waiting for.  Will pick this up and make the decision in January.

hacdias commented:
@lidel I added an option that will make `ProgressBar` use CSS animations. I also added an `overflow: hidden` to the parent `div` so that the colored part/progress does not overflow.

lidel reviewed (changes requested):
@hacdias I don't  believe https://github.com/ipfs/go-ipfs/issues/8591 or https://github.com/ipfs/go-ipfs/issues/8586 will happen this quarter,  let's not wait anymore. I think it is ok for us to ship it because    IPNS publishing should be in IPFS Desktop, and unlike the upstream kubo (go-ipfs), IPFS Desktop  has IPNS over pubsub enabled by default, so performance will be acceptable.

Mind resolving conflicts in this PR and pinging me when ready?  :pray:

SgtPooki commented:
@juliaxbow this is a great candidate for your review and help!

hacdias commented:
I tried a new way of rebasing and I guess I _messed up_ this PR. Sorry for that and won't happen again. The branch is fine, but this PR got closed and I can't re-open it. 

**Continue on #1973**.
